test/python/transpiler/test_elide_permutations.py

```diff
@@ -0,0 +1,350 @@
+        qc.h(0)
+        with qc.if_test((0, 0)):
+            qc.swap(0, 1)
+        qc.cx(0, 1)
+        res = self.swap_pass(qc)
+        self.assertEqual(res, qc)
+
+    def test_swap_if_else_block_with_outside_swap(self):
+        """Test swap elision only happens outside control flow."""
+        qc = QuantumCircuit(3, 3)
+        qc.h(0)
+        qc.swap(2, 0)
+        body = QuantumCircuit(2)
+        body.swap(0, 1)
+        if_else_op = IfElseOp((qc.clbits[0], 0), body)
+
+        qc.append(if_else_op, [0, 1])
+        qc.cx(0, 1)
+
+        expected = QuantumCircuit(3, 3)
+        expected.h(0)
+        expected.append(IfElseOp((expected.clbits[0], 0), body), [2, 1])
+        expected.cx(2, 1)
+
+        res = self.swap_pass(qc)
+        self.assertEqual(res, expected)
+
+    def test_swap_condition(self):
+        """Test swap elision doesn't touch conditioned swap."""
+        qc = QuantumCircuit(3, 3)
+        qc.h(0)
+        qc.swap(0, 1).c_if(qc.clbits[0], 0)
+        qc.cx(0, 1)
+        res = self.swap_pass(qc)
+        self.assertEqual(res, qc)
+
+    def test_unitary_equivalence_pass_manager(self):
+        """Test full transpile pipeline with pass preserves permutation for unitary equivalence."""
+        qc = QuantumCircuit(3)
+        qc.h(0)
+        qc.swap(0, 2)
+        qc.cx(0, 1)
+        qc.swap(1, 0)
+        qc.h(1)
+
+        expected = QuantumCircuit(3)
+        expected.h(0)
+        expected.cx(2, 1)
+        expected.h(2)
+
+        # First assert the pass works as expected
+        res = self.swap_pass(qc)
+        self.assertEqual(res, expected)
+
+        # with no layout
+        res = transpile(qc, optimization_level=3, seed_transpiler=42)
+        self.assertTrue(Operator.from_circuit(res).equiv(expected))
+        # With layout
+        res = transpile(
+            qc, coupling_map=CouplingMap.from_line(3), optimization_level=3, seed_transpiler=1234
+        )
+        self.assertTrue(Operator.from_circuit(res).equiv(expected))
+
+    def test_unitary_equivalence_with_elide_and_routing(self):
+        """Test full transpile pipeline with pass preserves permutation for unitary equivalence including
+        a larger example and a basis translation."""
+        qc = QuantumCircuit(5)
+        qc.h(0)
+        qc.swap(0, 2)
+        qc.cx(0, 1)
+        qc.swap(1, 0)
+        qc.cx(0, 1)
+        qc.cx(0, 2)
+        qc.cx(0, 3)
+        qc.cx(0, 4)
+        qc.h(1)
+
+        expected = QuantumCircuit(5)
+        expected.h(0)
+        expected.cx(2, 1)
+        expected.cx(1, 2)
+        expected.cx(1, 0)
+        expected.cx(1, 3)
+        expected.cx(1, 4)
+        expected.h(2)
+
+        # First assert the pass works as expected
+        res = self.swap_pass(qc)
+        self.assertEqual(res, expected)
+
+        # with no layout
+        res = transpile(qc, optimization_level=3, seed_transpiler=42)
+        self.assertTrue(Operator.from_circuit(res).equiv(expected))
```

Comment:
Since `Operator.from_circuit(res)` should take the initial and the final layouts into account, shouldn't we be checking (here and in other testcases) that `Operator.from_circuit(res)` is equivalent to the original circuit `qc`?

Reply:
Yeah, I was also wondering about this, thanks for pointing it out! :-) My understanding is that swap gates 'in the middle' of `qc` change the final layout of `qc` with respect to the 'canonical' layout generated by `Operator.from_circuit(res)`. However, this change is not recorded in the `_layout` field of `qc`.

 So, `Operator.from_circuit(res).equiv(qc)` does not need to be true while `Operator.from_circuit(res).equiv(Operator.from_circuit(qc, final_layout=fl))` would be true for an appropriate choice of a final layout `fl`. I found it more time-efficient to just specify an expected quantum circuit than to figure out the correct `final_layout` for each example. If you think this is important to validate this nevertheless I can figure out the correct `fl` for each example. :-)